File: 10/solution.js
const fs = require('fs')
const {splitIntoChunks} = require("../helpers");

function readInput() {
    return fs.readFileSync('./day-10/input.txt', 'utf8')
        .split('\n')
        .map(line => {
            const [command, num] = line.split(' ')
            return command === 'addx' ? Number(num) : 0 // would not work with "addx 0" but oh well
        })
}

const input = readInput()

function resolveValuesForCycles(inputNums, numsByCycle, nextValue) {
    if (!inputNums.length) return numsByCycle
    const nextInputNum = inputNums.shift()
    const updatedNums = nextInputNum ? numsByCycle.concat([nextValue, nextValue]) : numsByCycle.concat(nextValue)
    return resolveValuesForCycles(inputNums, updatedNums, nextValue + nextInputNum)
}

function signalStrengthSum(cycleToValue) {
    let total = 0
    for (let i = 20; i < cycleToValue.length; i += 40) {
        total += cycleToValue[i] * i
    }
    return total
}

function solvePart1() {
    const cycleToValue = resolveValuesForCycles(input.slice(), [null], 1)
    return signalStrengthSum(cycleToValue)
}

function solvePart2() {
    const cycleToValue = resolveValuesForCycles(input.slice(), [null], 1)
    const pixels = Array(240).fill('.')
    for (let i = 0; i < cycleToValue.length; i++) {
        const spritePositions = getSpritePositions(cycleToValue[i + 1])
        if (spritePositions.includes(i % 40)) pixels[i] = '#'
    }
    return printResponse(pixels)
}

function getSpritePositions(middle) {
    return [middle - 1, middle, middle + 1]
}

function printResponse(pixels) {
    return splitIntoChunks(pixels, 40).map(chunk => chunk.join('')).join('\n')
}

console.log(solvePart1() === 14780)
console.log(solvePart2())

File: 10/solution.scala
import scala.io.Source

object Solution extends App {
  val lines = Source.fromFile("./input.txt").getLines.map(_.split(" ").toList).toList

  println(solvePart1(lines) == 14780)
  println(solvePart2(lines))

  def solvePart1(lines: List[List[String]]): Int = {
    val valuesByCycle = resolveValuesForCycles(lines)
    signalStrengthSum(valuesByCycle)
  }

  def solvePart2(lines: List[List[String]]): String = {
    val valuesByCycle = resolveValuesForCycles(lines)
    val pixels = valuesByCycle.zipWithIndex.foldLeft(List.empty) { case (pixels, (value, index)) =>
      if (spritePositions(value).contains(index % 40)) pixels :+ "#" else pixels :+ "."
    }
    pixels.grouped(40).map(_.mkString).mkString("\n")
  }

  def spritePositions(value: Int): List[Int] = List(value - 1, value, value + 1)

  def signalStrengthSum(numsByCycle: List[Int]): Int = {
    (19 to 239 by 40).map(index => (index + 1) * numsByCycle(index)).sum
  }

  def resolveValuesForCycles(lines: List[List[String]]): List[Int] = {
    lines.foldLeft((List.empty[Int], 1)) { case ((valuesByCycle, lastValue), line) =>
      val newValuesByCycle = valuesByCycle :+ lastValue
      line match {
        case command :: numFromInput :: Nil => (newValuesByCycle :+ lastValue, lastValue + numFromInput.toInt)
        case _ => (newValuesByCycle, lastValue)
      }
    }
  }._1
}